BLE Indoor Positioning/src/main/java/com/nexenio/bleindoorpositioning/location/provider/DeviceLocationPredictor.java

```diff
@@ -0,0 +1,137 @@
+                        pastDeviceLocations.get(i).getLongitude(),
+                        pastDeviceLocations.get(i + 1).getLatitude(),
+                        pastDeviceLocations.get(i + 1).getLongitude());
+            }
+        }
+
+        //in meter
+        double meanDistance = sumDistances / pastDeviceLocations.size();
+
+        // TODO what if no movement --> Stay-point detection
+        if (meanDistance > 0.6) {
+            isMoving = true;
+        }
+
+        // calculate bearing based on last two known locations
+        double bearing = 0;
+        if (pastDeviceLocations.size() > 1) {
+            for (int i = pastDeviceLocations.size() - 1; i >= pastDeviceLocations.size() - 1; i--) {
+                bearing = calculateBearing(
+                        pastDeviceLocations.get(i - 1).getLatitude(),
+                        pastDeviceLocations.get(i - 1).getLongitude(),
+                        pastDeviceLocations.get(i).getLatitude(),
+                        pastDeviceLocations.get(i).getLongitude());
+            }
+        }
+
+        // TODO check  different angle calculations - getRotationAngleInDegrees & calculateBearing
+        // double angle = Location.getRotationAngleInDegrees(pastDeviceLocations.get(i - 1), pastDeviceLocations.get(i));
+        // System.out.println("bearing: " + bearing + " | angle:" + angle + " | diff: " + (bearing - angle));
+
+        // TODO calculate confidence of predicting --> is it reasonable to predict new location
+        // compare large window angle with small window angle
+
+        // set initial location
+        if (pastDeviceLocations.size() == 0) {
+            predictedLocation = deviceCenter;
+        } else {
+            predictedLocation = calculateNextLocation(pastDeviceLocations.get(pastDeviceLocations.size() - 1).getLatitude(),
+                    pastDeviceLocations.get(pastDeviceLocations.size() - 1).getLongitude(),
+                    meanDistance / 1000, bearing);
+        }
+        setPredictedLocation(predictedLocation);
+    }
+
+    // TODO compare prediction with next location
+
+    /**
+     * @return the angle
+     */
+
+    static double calculateBearing(double fromLatitude, double fromLongitude, double toLatitude, double toLongitude) {
+        fromLatitude = Math.toRadians(fromLatitude);
+        toLatitude = Math.toRadians(toLatitude);
+        double longitudeDifference = Math.toRadians(toLongitude - fromLongitude);
+        double y = Math.sin(longitudeDifference) * Math.cos(toLatitude);
+        double x = (Math.cos(fromLatitude) * Math.sin(toLatitude)) - (Math.sin(fromLatitude) * Math.cos(toLatitude) * Math.cos(longitudeDifference));
+        return (Math.toDegrees(Math.atan2(y, x)) + 360) % 360;
+    }
+
+    /**
+     * @param distance in km
+     * @param bearing  in degrees
+     * @return Location
+     */
+    private static Location calculateNextLocation(double fromLatitude, double fromLongitude, double distance, double bearing) {
```

Comment:
This should be a convenience method in `Location` that accepts distance and bearing